rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1450 @@
+  }
+  ```
+  - This violates the guideline of the indentation representing the expression structure, and thus reduces readability.
+
+### Function declaration
+
+- The body of the function must not be indented relative to its first arguments.
+- A small number of ("simple") identifier arguments can be written onto the same line.
+  - Otherwise they're each on their own line.
+  - The body may get absorbed here
+- Attribute set arguments must always start on a new line and they must not be mixed with identifier arguments.
+  - If they have few attributes, the argument may be written on a single line
+  - Otherwise each attribute must be on its own line with indentation, followed by a trailing comma.
+- Due to structural similarity and for consistency reasons, attribute set arguments with a default value follow the same rules as [bindings](#bindings).
+
+**Examples**
+
+```nix
+#1
+name: value: name ++ value
+
+#2 absorption
+name: value: ''
+  ${name} = ${value};
+''
+
+#3
+name: value:
+name
+++ value
+++ more stuff making the line too long
+
+#4
+{ pkgs }: pkgs.hello
+
+#5
+args@{
+  some,
+  argument,
+  default ? value,
+  ...
+}:
+{
+  # body
+}
+
+#6
+{ pkgs }:
+name: value: 
+{
+  # body
+}
+
+#7: These would be over the line length limit on a single line
+aaaa:
+bbbb:
+cccc: 
+dddd:
+eeee:
+null
+
+#8: @ patterns can also come afterwards
+{ pkgs }@args: pkgs.hello
+```
+
+**Alternatives**
+
+- Have leading commas for parameters in attribute set arguments, like currently done in Nixpkgs.
+
+  - This makes attribute set arguments less likely to be confused with lists.
+  - It's easier to see where arguments start and end.
+  ```nix
+  { some
+  , arg
+  }:
+  
+  args@{
+    some
+  , argument
+    # Single line comment
+  , commentedArgument
+  , # Comment on the value
+    # multiline comment
+    default ? value
+  , ...
+  }:
+  # ...
+  ```
+
+  Problems with this alternative:
+  - Moving items around with this style may require editing lines.
+  - Inconsistent with the [expression expansion guidelines](#expansion-of-expressions), which disallows forms like `{ some`; `some` should start on a new line instead.
+  - This does not work well with leading `@` bindings.
+  - It's unclear whether comments belong to the next or the previous argument.
+  - The leading comma style was a lesser-evil workaround for the lack of trailing commas in the Nix language. Now that the language has this feature, there is no reason to keep it that way anymore.
+
+### Operators
+
+From the [list of operators](https://nixos.org/manual/nix/stable/language/operators.html#operators), this section focuses on binary operators.
+Function application and attribute selection are not treated as an "operator" in the sense of this section, see [function application](#function-application) instead.
+
+#### Non-chainable operators
+
+Operators with no associativity are non-chainable.
+Each invokation will always have exactly one left-hand side and one right-hand side.
```

Comment:
```suggestion
Each invocation will always have exactly one left-hand side and one right-hand side.
```

Reply:
Fixed in 1e5914d2a214f5b3bf1e3a1279fd683cb7b18c2b